[GTK][WPE] Use the IPC connection to send/receive credentials

carlosgcampos opened the pull request:
#### 79e0f1258ac535a59aead696ac57444fb72bd3b6
<pre>
[GTK][WPE] Use the IPC connection to send/receive credentials
<a href="https://bugs.webkit.org/show_bug.cgi?id=297846">https://bugs.webkit.org/show_bug.cgi?id=297846</a>

Reviewed by Michael Catanzaro.

We currently use a dedicated socket pair connection for that, but we
could just use the IPC connection. Also, we only need to do that when
spawning sandboxed web processes, for other process types or when
sandbox is disabled we already have the pid, so we don&apos;t need to slow
down the process launch.

Canonical link: <a href="https://commits.webkit.org/299151@main">https://commits.webkit.org/299151@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a8916225d426cbe9c76c09bcacc97519096a326

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124188 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70074 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38409 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46298 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89574 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59177 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120990 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30593 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70066 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29661 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67853 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100016 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24138 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127269 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33866 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98246 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102067 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98032 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43434 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21422 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41377 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18811 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44813 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50487 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44273 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47618 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45962 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -610,73 +595,35 @@ std::optional<Connection::ConnectionIdentifierPair> Connection::createConnection
+
+    pid_t processID = g_credentials_get_unix_pid(credentials.get(), &error.outPtr());
+    if (error)
+        g_error("Connection: failed to get pid from credentials: %s\n", error->message);
```
```suggestion
        g_error("Connection: failed to get pid from credentials: %s", error->message);
```

mcatanzaro commented on Source/WebKit/Shared/AuxiliaryProcess.cpp:
```diff
@@ -132,8 +135,14 @@ void AuxiliaryProcess::initializeProcessName(const AuxiliaryProcessInitializatio
+void AuxiliaryProcess::initializeConnection(IPC::Connection* connection)
 {
+#if OS(LINUX) && USE(GLIB)
+    if (isInsideFlatpak())
```
Two thoughts here:

1. We need to do this for bubblewrap sandbox as well, not only flatpak sandbox. I would at least add a comment here to explain why the code is correct, because it looks like it's not. I see that when we are the web process, this check will return true because the bubblewrap sandbox _looks_ like a Flatpak sandbox, i.e. `isInsideFlatpak` just checks for Flatpak in the UI process, but in the web process it additional checks for WebKit's own sandbox. Maybe we should follow up on this because it's quite confusing.
2. Sending credentials is not necessary if we are the parent process rather than the child process. But I suppose it's also harmless, and not worth adding code to avoid doing so?

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -245,28 +233,33 @@ void ProcessLauncher::launchProcess()
-
-        m_socketMonitor.stop();
+    if (isInsideFlatpak() || (sandboxEnabled && shouldUseBubblewrap())) {
+        GRefPtr<GSocket> pidSocket = adoptGRef(g_socket_new_from_fd(webkitSocketPair.server.release(), &error.outPtr()));
```
Previously we used a separate pid socket here, but now it's just the normal connection socket, so this name doesn't make sense anymore. Maybe you could call it just `socket` instead of `pidSocket`.

mcatanzaro reviewed (approved):
Nice cleanup. This is what I should have done originally. Looks like I made things more  complicated than necessary.

carlosgcampos committed:
[GTK][WPE] Use the IPC connection to send/receive credentials
https://bugs.webkit.org/show_bug.cgi?id=297846

Reviewed by Michael Catanzaro.

We currently use a dedicated socket pair connection for that, but we
could just use the IPC connection. Also, we only need to do that when
spawning sandboxed web processes, for other process types or when
sandbox is disabled we already have the pid, so we don't need to slow
down the process launch.

Canonical link: https://commits.webkit.org/299151@main
